stopify-continuations/src/runtime/abstractRuntime.ts

```diff
@@ -132,27 +144,69 @@ export abstract class Runtime {
     };
   }
 
-  // TODO(rachit): Document this.
+  // Top level function that runs a given program. Handles capture and restore
+  // events that may be emitted by the program.
   runtime(body: () => any): any {
-    while (true) {
-      const result = this.abstractRun(body)
-      if (result.type === 'normal') {
-        assert(this.mode, 'executing completed in restore mode');
-        return;
+
+    while(true) {
+      const result = this.abstractRun(body);
+
+      if (result.type === 'normal' || result.type === 'exception') {
+        if (this.savedStack.length > 0) {
+          const exception = result.type === 'exception';
+          const ss = this.savedStack;
+
+          const restarter: KFrameTop = this.topK(() => {
+            if (exception) throw result.value;
+            else return result.value;
+          });
+
+          this.stack = [restarter];
+
+          for (let i = 0; i < this.restoreFrames && ss.length - 1 - i >= 0; i++) {
+            this.stack.push(ss.pop()!);
+          }
+          body = () => {
+            this.mode = false;
+            return this.stack[this.stack.length - 1].f()
+          }
+        }
+        else if(result.type === 'normal') {
+          assert(this.mode, 'execution completed in restore mode')
+          return result.value;
+        }
+        else if(result.type === 'exception') {
+          assert(this.mode, `execution completed in restore mode, error was: ${result.value}`)
+          throw result.value;
+        }
+        else {
+          unreachable()
+        }
       }
       else if (result.type === 'capture') {
-        body = () => result.f.call(global, this.makeCont(result.stack));
+        const s = result.stack;
+        body = () => {
+          for(let i = s.length - 1; i >= this.restoreFrames; i -= 1) {
+            this.savedStack.push(s.pop()!);
+          }
+
+          const savedStack = this.savedStack;
```

Comment:
I'm confused about what is happening here. You are pushing onto `this.savedStack` is suffix of `result.stack`. You then capture `this.savedStack` in the continuation.

I expected that when capturing, the stack saved in the continuation would be `this.savedStack.append(result.stack)`.

Reply:
There need to be two stacks, one that is restored onto the JS stack, and one that is saved.
I’m saving the suffix of the stack frame and passing the rest of it to the continuation.

Reply:
Ah, I see now. makeCont takes two stack arguments. I was expecting it to take a single one and figured you would append the two stacks before calling makeCont. Any reason to not do that? It may make the representation of continuations slightly simpler.

Reply:
Sounds cleaner. I can do that. It will also make the makeCont for eager cleaner.

Reply:
Great. This is going to be amazing.

Reply:
Done